Fixes #22236 - Adds breadcrumbs to foreman

amirfefer opened the pull request:
![compute resources](https://user-images.githubusercontent.com/11807069/34943288-bb9f72c8-fa03-11e7-8b06-30ce9dbb1efc.gif)

amirfefer committed:
Fixes #22236 - Adds breadcrumbs to foreman UI

timogoebel commented:
@amirfefer: Note, that we already have breadrumbs for ssh keys:
https://github.com/theforeman/foreman/blob/00183e17ad70876385106a2e1979c5b52012f69d/app/views/ssh_keys/new.html.erb#L3-L7

waldenraines commented:
I get the following on the new compute resources page:

```
15:09:19 rails.1   | 2018-01-11T15:09:19 c35dcf0e [app] [F] ActionView::Template::Error (undefined method `edit_compute_resources_path' for #<#<Class:0x007f8cb62f67b0>:0x007f8cb993c720>):
15:09:19 rails.1   | 2018-01-11T15:09:19 c35dcf0e [app] [F]     1: <%= breadcrumbs('name', @compute_resource ) %>
15:09:19 rails.1   |  |     2: <%= form_for @compute_resource, :html => {:data => {:id => @compute_resource.try(:id)}} do |f| %>
15:09:19 rails.1   |  |     3:   <%= base_errors_for @compute_resource %>
15:09:19 rails.1   |  |     4:   <ul class="nav nav-tabs" data-tabs="tabs">
15:09:19 rails.1   | 2018-01-11T15:09:19 c35dcf0e [app] [F]   
15:09:19 rails.1   | 2018-01-11T15:09:19 c35dcf0e [app] [F] app/helpers/layout_helper.rb:25:in `block in breadcrumbs'
15:09:19 rails.1   |  | app/helpers/layout_helper.rb:20:in `breadcrumbs'
15:09:19 rails.1   |  | app/views/compute_resources/_form.html.erb:1:in `_8cff77e70f47deae4d41b9454e8b9ec0'
15:09:19 rails.1   |  | app/views/compute_resources/new.html.erb:3:in `_1ec78fed1c788d4c96ad944eee20c78b'
15:09:19 rails.1   |  | app/controllers/concerns/application_shared.rb:15:in `set_timezone'
15:09:19 rails.1   |  | app/models/concerns/foreman/thread_session.rb:32:in `clear_thread'
15:09:19 rails.1   |  | app/controllers/concerns/foreman/controller/topbar_sweeper.rb:12:in `set_topbar_sweeper_controller'
15:09:19 rails.1   |  | lib/middleware/catch_json_parse_errors.rb:8:in `call'
15:09:19 rails.1   |  | lib/middleware/session_safe_logging.rb:17:in `call'
15:09:19 rails.1   |  | lib/middleware/tagged_logging.rb:18:in `call'
```

ohadlevy commented on webpack/assets/javascripts/react_app/components/Breadcrumb/index.js:
```diff
@@ -0,0 +1,11 @@
+import { Breadcrumb } from 'patternfly-react';
+
+const ForemanBreadcrumb = ({ data }) => <Breadcrumb>
+        {data.menu.map((item, index) =>
```
please configure prettier

amirfefer commented:
@timogoebel :+1:  I've changed it to react component
@waldenraines: I've fixed this regression
@ohadlevy: I'm using foreman's content layout to render `breadcrumbs`

amirfefer commented:
Rebased

amirfefer commented:
rebased + test failures should be fixed

timogoebel commented:
@amirfefer: Patternfly recommends the breadcrumbs to be [placed underneath the navbar](http://www.patternfly.org/pattern-library/navigation/breadcrumbs/). Why did you choose a different approach?

amirfefer commented:
@timogoebel  thanks for mentioning  that
patternfly placement for it looks a bit odd to me:

![breadcrumbs-1](https://user-images.githubusercontent.com/11807069/35096054-68eac1de-fc53-11e7-844b-dcfd84ed938f.png)
VS

![breadcrumbs-2](https://user-images.githubusercontent.com/11807069/35096063-6e76573a-fc53-11e7-901e-25805cff4544.png)

ohadlevy commented:
@rohoover  should we consider combining the header the breadcrums ? (note that some headers have images / icons next to the text).

ohadlevy reviewed (changes requested):
@amirfefer thanks!

Can you please add a generic test (similar to pagination) that goes over all pages to ensure they actually have pagination rendered correctly?

ohadlevy commented on app/views/layouts/_application_content.html.erb:
```diff
@@ -15,6 +15,15 @@
+        <% if content_for?(:breadcrumbs) %>
+          <%= yield(:breadcrumbs) %>
+        <% else %>
+          <%= mount_breadcrumbs(get_resource) if !searchable? && get_resource %>
```
Should we really use searchable as a way to define breadcrums ? even if its the same condition I would prefer to have a dedicated method.

I'm also not sure that if there is a search component we should not show breadcrums.. ?

tstrachota replied:
+1 for having a dedicated method.
I have no strong opinion about checking the `!searchable?` we can tune it later if it doesn't work 100%.

sharvit commented:
https://gist.github.com/sharvit/41d73803d5e15e5348c5baa7be758b79

Rohoover commented:
No I wouldn't combine them. 

The reason it's below the nav is because that is more universal with all the different implementations of content. In the following link you can see the different iterations of design that utilize the breadcrumb:
http://www.patternfly.org/pattern-library/navigation/breadcrumbs/

It just so happens that we do not have a the other visual elements, such as the grey background, that help visually arrange the items. 

I would suggest grouping the title and content a little closer.

timogoebel commented:
> It just so happens that we do not have a the other visual elements, such as the grey background,

Is this planned? The cards e.g. don't look any good on white background unless they are accented.

Rohoover commented:
@timogoebel 
My mistake, I noticed further down in the documentation there is a breadcrumb design combined with title.

ohadlevy commented:
> I noticed further down in the documentation there is a breadcrumb design combined with title.

@Rohoover so in that case do you suggest to remove the title as it is today and combine it with the breadcrumbs? thanks!

Rohoover commented:
@ohadlevy 
Yes and it would be preferred if we could move the Katello portions to this as well so it's consistent (@waldenraines).

timogoebel commented:
@Rohoover: What about the grey background. Should we use that as well for normal pages or just with a "cards page". Although it's really off-topic, I'd love to see how the host view page would look using the cards layout.

Rohoover commented:
@timogoebel 
I think that would be a good idea. PF calls for it anyways and it plays up nicely against many elements including the cards.

waldenraines reviewed (approved):
This works for me everywhere I tested.

tstrachota commented:
@amirfefer could you rebase, please?

tstrachota commented on app/views/compute_resources/show.html.erb:
```diff
@@ -5,7 +5,6 @@
                                             :title => _("Associate VMs to Foreman hosts"), :method => :put, :class=>"btn btn-default"),
 (display_link_if_authorized(_('Refresh Cache'), hash_for_refresh_cache_compute_resource_path(@compute_resource).merge(:auth_object => @compute_resource), :method => :put, :class => "btn btn-default") if @compute_resource.respond_to?(:refresh_cache) && @compute_resource.caching_enabled),
    link_to_if_authorized(_('Edit'), hash_for_edit_compute_resource_path(@compute_resource).merge(:auth_object => @compute_resource), :class => "btn btn-default")  %>
-
```
This seems unrelated.

tstrachota commented on app/helpers/layout_helper.rb:
```diff
@@ -16,6 +16,24 @@ def search_bar(*elements)
+  def mount_breadcrumbs(obj, &block)
+    label = obj.try(:to_label) || obj.try(:to_s) unless obj.new_record?
+    default_menu = [{caption: _(controller_name.humanize), url: try("#{controller_name}_path")},
+                    {caption: [_(action_name.humanize), label].join(' '), url: '#' }].compact unless block_given?
```
I'm not sure about showing the action in breadcrumbs. It's not consistent with katello and I don't think it adds much value. It also makes the breadcrumbs to change when you switch for example from "user edit" `Users >> Edit John` to "add breadcrumbs" `Users >> John >> Add SSH Key`.
I'd prefer displaying just the record's name without the action prefix.

amirfefer replied:
I removed the action referring  from the breadcrumbs, now its combined with the title itself.

tstrachota reviewed (changes requested):
It works well. I added couple of comments inline. I'd prefer removing the action name from breadcrumbs for the sake of consistency.

tstrachota commented:
@amirfefer it still needs a rebase

amirfefer commented:
@tstrachota 
Thanks for your review.
I've combined the header title with the breadcrumbs, as it described in patternfly design.

tstrachota commented:
@amirfefer I think that @Rohoover was actually against combining title and breadcrumbs in our case (see https://github.com/theforeman/foreman/pull/5167#issuecomment-359442794). And I agree with that. It's inconsistent with how Katello handles breadcrumbs:

![katello_product](https://user-images.githubusercontent.com/1186631/36254664-2a67a85a-124c-11e8-9631-b0eab0dd94d2.png)

It looks quite odd with two font sizes in the breadcrumbs:

![user_edit](https://user-images.githubusercontent.com/1186631/36254656-2325b1cc-124c-11e8-81d5-d81c8a2fac73.png)
![user_ssh_key](https://user-images.githubusercontent.com/1186631/36254659-26b77424-124c-11e8-96fd-547e1d5ff2b7.png)
![host_detail](https://user-images.githubusercontent.com/1186631/36254663-28a2e642-124c-11e8-99f6-096464cc50de.png)

It also doesn't solve the issue with action names in breadcrumbs (https://github.com/theforeman/foreman/pull/5167#discussion_r166246945). They're still there. So I think the previous layout was better. It just needed removal of the action names.

**To be consistent with katello and in line with good practices I believe the breadcrumbs should follow the pattern:**

_on detail or edit pages:_ 
`[resource_name_plural]` > `[record_name]` 
e.g. `Users > admin`

_on new record pages:_ 
`[resource_name_plural]` > `New [resource_name_singular]` 
e.g. `Users > New user`

_on listing of nested resoruces:_ 
`[resource_name_plural]` > `[record_name]` > `[nested_resource_name_plural]` 
e.g. `Compute resources > libvirt > Images`

Rohoover reviewed (commented):
@tstrachota 

Yes initially I was until I saw the additional patternFly documentation on it. It was also brought to my attention that other areas are starting to use this approach, so we'll be a bit ahead of the curve.

I share your concerns about the text size. I'm going to bring this to the PF teams attention.

waldenraines commented:
> Yes initially I was until I saw the additional patternFly documentation on it. It was also brought to my attention that other areas are starting to use this approach, so we'll be a bit ahead of the curve.
> 
> I share your concerns about the text size. I'm going to bring this to the PF teams attention.

Wait, so we're saying that we don't want a title on the page and instead just the breadcrumbs?  I personally don't like the way this looks.

tstrachota commented:
Ok, I had to read through the whole thread again. Now I understand it, @Rohoover.

I agree with @waldenraines, I don't like the current look. Using gray background and white breadcrumbs bar would probably make it better, but I'm afraid we'd have to sort out title action buttons placement.

I vote for iterative approach, smaller steps at a time:
Let's get the breadcrumbs in with titles now and we can always improve it later.

waldenraines commented:
> I vote for iterative approach, smaller steps at a time:
> Let's get the breadcrumbs in with titles now and we can always improve it later.

+1.  We already have this style of breadcrumbs in katello.  If we change what we do here then we have to redo katello.

Rohoover reviewed (commented):
@waldenraines @tstrachota 

+1 to an iterative approach.

Some of the additional documentation on it is here (for anyone who hasn't seen).
http://www.patternfly.org/pattern-library/navigation/breadcrumbs/#design

All the designs showf an ideal situation with a card layout that offers a grey background to contrast the white breadcrumb area, this is obviously not our typical visual use case. Maybe add a grey line to separate it out.

timogoebel commented:
I personally like an iterative approach in general, but I think we should make sure, that when we add new groundbreaking functionality (i.e. breadcrumbs) we go all in so users don't have to learn the feature twice. Why do we have patternfly standards, if we deviate when it's more convenient for us?
The document clearly describes how breadcrumbs should look. And I do think, it looks good and integrates well with the rest of the layout.

http://www.patternfly.org/pattern-library/navigation/breadcrumbs/img/Breadcrumbs-02.png

> We already have this style of breadcrumbs in katello. If we change what we do here then we have to redo katello.

I think core should set the standard here, even if that means that plugins need to adapt.

waldenraines commented:
> I personally like an iterative approach in general, but I think we should make sure, that when we add new groundbreaking functionality (i.e. breadcrumbs) we go all in so users don't have to learn the feature twice. Why do we have patternfly standards, if we deviate when it's more convenient for us?

This is a good point and I think you just convinced me.  It will be a little bit more work but I do think it makes sense to adhere to patternfly designs when we have them.  I don't think this design was available when the breadcrumbs in katello were implemented but now that it is we should spend the effort to adhere to the design.

Rohoover reviewed (commented):
@waldenraines @timogoebel 
Consistency between Katello/Foreman is of utmost importance, which I think we have consensus on.

Basically even I don't always agree with PF, but better adhere to consistency and standards, and push the upstream (PF) to change than diverge "just because we don't like it".

Hope that makes sense.

waldenraines commented:
See also http://projects.theforeman.org/issues/22603 - Make breadcrumbs and resource switcher conform to patternfly design (Katello)

amirfefer commented:
I opened a [PR](https://github.com/patternfly/patternfly-react/pull/233) in `patternfly-react` which implements breadcrumbs combined with page title.
@tstrachota page titles are already including the specific  action, why would you like to remove this?
IMHO if we going to use breadcrumbs combined with page header, we should keep the current behavior - which means keep the action title.

tstrachota commented:
Since breadcrumbs should give you a hint where you are in the app structure, I consider "Edit user" or "Edit template" insufficient. I know that it's how the titles are implemented now, but imo it deserves improvement.

I'd expect `Users > Edit admin` or just `Users > admin` so that it's clearer what you're editing.
"Add user" is fine, since there's probably nothing better we could display (maybe just "Add").

For the resources that have show pages (proxy, host) I'd expect the breadcumbs to be `Users > admin` for the detail and `Users > admin > edit` for the editing form.

amirfefer commented:
@tstrachota, Thanks however,  I think that breadcrumbs combined with page header should give you a full reference where you are, not just a hint.
@Rohoover what is your point of view about that?

sharvit commented:
I agree with @amirfefer, I think it makes sense to show Breadcrumbs as a title that gives you a full reference to where you are.

Rohoover reviewed (commented):
@amirfefer 
Can you provide me a current example of what is shown and what the new implementation would be? Sorry, I'm having trouble understanding exactly what you all are referring to.

tstrachota commented:
> page header should give you a full reference where you are

I should have phrased it better, we're not in contradiction on this. That's why I'd like to have the record names in breadcrumbs for detail and edit pages.

waldenraines commented:
> Can you provide me a current example of what is shown and what the new implementation would be? Sorry, I'm having trouble understanding exactly what you all are referring to.

@Rohoover see Tomas's [comment here](https://github.com/theforeman/foreman/pull/5167#issuecomment-366917208) for an example.

Rohoover reviewed (commented):
@tstrachota @waldenraines 

Thanks. Yes, if I'm understanding correctly, then I agree with you.

amirfefer commented:
@tstrachota 
> For the resources that have show pages (proxy, host) I'd expect the breadcumbs to be Users > admin for the detail and Users > admin > edit for the editing form.

So in that case  if you click on `admin`, will you be directed to the admin's `show` page? 
however, your last page probably was index.
breadcrumbs should present location hierarchy, I don't see it in `Users > admin (show) > edit (admin)

If I have understood correctly, it should be like this:
`Users > Edit admin` + switcher button in order to switch the resource (i.e from `edit admin` to `edit John Doe`)

amirfefer commented:
I'm using `breadcrumbs combined with page title` patternfly-react implementation
@ohadlevy - it's problematic to create  a generic test for each and every page (like pagination you mentioned) 
pagination renders in the server, but in this case the rendering occurs  in the client, which means we need a fast and stable javascript driver, after some tests  I made with `Poltergeist` it would take time.
I don't think we should test react mounted components via ruby integration test with the today's configuration

mmoll commented:
[test foreman]
[test katello]

sharvit commented on package.json:
```diff
@@ -73,7 +73,7 @@
     "multiselect": "~0.9.12",
     "patternfly": "^3.31.2",
-    "patternfly-react": "1.11.0",
+    "patternfly-react": "^1.12.0",
```
Can you pin it?
We decided to use pin versions of `patternfly-react` until we will migrate to shallow snapshots. It just break to much on Travis.

sharvit reviewed (changes requested):
Looks good to me!
can you pin `patternfly-react`?

ohadlevy commented on app/helpers/layout_helper.rb:
```diff
@@ -16,6 +16,21 @@ def search_bar(*elements)
+                    {caption: @page_header, url: '#' }].compact unless block_given?
+
+    mount_react_component("Breadcrumb", "#breadcrumb",
+      { menu:  default_menu || yield }.to_json)
```
nitpick: extra whitespace after menu

ohadlevy commented on app/views/ssh_keys/new.html.erb:
```diff
@@ -1,9 +1,7 @@
+<%= breadcrumbs do %>
+  <% [{caption: _("Users"), url: (url_for(users_path) if authorized_for(hash_for_users_path))},
+      {caption: @user.name, url: (edit_user_path(@user) if authorized_for(hash_for_edit_user_path(@user)))},
+      {caption: _('Add SSH Key'), url: '#'}] %>
```
are you required to add the url: options even if you dont use it?

ohadlevy commented on webpack/assets/javascripts/react_app/components/Breadcrumb/Breadcrumbs.fixtures.js:
```diff
@@ -0,0 +1,13 @@
+    },
+    {
+      caption: 'child',
+      url: '#',
```
just double checking I understand this correctly - this is not really required? (for the child)

amirfefer replied:
right :+1:

ohadlevy commented on app/views/layouts/_application_content.html.erb:
```diff
@@ -1,21 +1,33 @@
+          <% unless @page_header.blank? %>
+            <div class="row form-group">
+              <h1 class="col-md-8">
+                <%=h @page_header %>
```
I find the the header in the wrong place / size? is that intentional?

![image](https://user-images.githubusercontent.com/20167/37254140-1635338e-2543-11e8-8afc-f07ac7ea1f1c.png)


vs 

![image](https://user-images.githubusercontent.com/20167/37254143-1dc3d1a0-2543-11e8-98ba-2c2df039673a.png)

tstrachota replied:
I think that we either need to align the title with the rest of the elements or add the white row on gray background as patternfly design suggests (http://www.patternfly.org/pattern-library/navigation/breadcrumbs/#design).
Having it as it is now would IMHO visually be a step back.

Rohoover replied:
Unfortunately the PF designs are only showing cards which are typically on grey backgrounds. There is however a note in the documentation that a grey horizontal line can be added:

"Divider (optional): A horizontal divider may be present below the breadcrumbs depending on the page layout used inside an application. If this format is used, it should be used consistently throughout the application."

tstrachota replied:
I think that we either need to align the title with the rest of the elements or add the white row on gray background as patternfly design suggests (http://www.patternfly.org/pattern-library/navigation/breadcrumbs/#design).
Having it as it is now would IMHO visually be a step back.

ohadlevy commented on app/views/layouts/_application_content.html.erb:
```diff
@@ -1,21 +1,33 @@
+          <% if content_for?(:breadcrumbs) %>
+            <%= yield(:breadcrumbs) %>
+          <% else %>
+            <%= mount_breadcrumbs %>
```
any idea why smart proxies have the wrong label? 
![image](https://user-images.githubusercontent.com/20167/37254214-be8b0e46-2543-11e8-9c6d-0111c91082e0.png)

ohadlevy replied:
is it possible to add hierarchy for roles / filters?
e.g.
```ruby
breadcrumbs do
  [
   {caption: _("Roles"), url: (url_for(roles_path) if authorized_for(hash_for_roles_path))},
   {caption: @role.name, url: (url_for(@role) if authorized_for(hash_for_edit_role_path(@role)))},
  {caption: _('Filters')}
  ]
end

```

ohadlevy replied:
when clicking on back (e.g. roles page) - the breadcrumbs / title disappear.
![image](https://user-images.githubusercontent.com/20167/37254247-11c7ce28-2544-11e8-9cc2-148f68d36e09.png)

tstrachota replied:
@ohadlevy I found this bug also on develop: http://projects.theforeman.org/issues/22901
Serious functionality issue, but not related to this PR.

ohadlevy commented on app/views/layouts/_application_content.html.erb:
```diff
@@ -1,21 +1,33 @@
+          <% if content_for?(:breadcrumbs) %>
+            <%= yield(:breadcrumbs) %>
+          <% else %>
+            <%= mount_breadcrumbs %>
```
can you please also handle pages such as facts, reports, audits etc

ohadlevy reviewed (changes requested):
overall looks good! thanks

a few questions:

tstrachota commented:
There's inconsistency in breadcrumbs for edit pages.

Editing user 'admin':
![edit_user_h](https://user-images.githubusercontent.com/1186631/37344355-769cd050-26ca-11e8-898f-abc15f1c7d2f.png)
Displays name of the resource.

Editing hostgroup 'Provisioning' and host 'clint-unrue':
![edit_hostgroup_h](https://user-images.githubusercontent.com/1186631/37344350-743de7d6-26ca-11e8-93b8-eadc67658306.png)
![host_edit_h](https://user-images.githubusercontent.com/1186631/37344363-78a0f2dc-26ca-11e8-8f56-60c579e4e6b7.png)
Displays name of the record.

I spotted this at multiple places. Personally I find the latter version better - it's clear what record you're editing.

tstrachota commented:
Breadcrumbs for audits seem quite confusing to me. Especially when edit breadcrumbs are `Edit [record.name]` and not `[record.name] >> edit` I had a feeling that I'm on some action page when I saw them for the first time. Past tense could help.

But I'm ok with fixing it later in a different PR. The audits page is probably going to be changed anyway.

![audit_detail_h](https://user-images.githubusercontent.com/1186631/37344893-ed48df4a-26cb-11e8-94f4-09f056d67890.png)
![audit_detail_h_2](https://user-images.githubusercontent.com/1186631/37344897-eed9680c-26cb-11e8-87e1-797868d297bb.png)
![audit_detail_h_3](https://user-images.githubusercontent.com/1186631/37344899-f04769dc-26cb-11e8-905a-f4b94940712c.png)

tstrachota commented:
I found some issues at compute profile pages.

The compute profile detail page shows "Edit" in the breadcrumb (nitpick).
![compute_profile_detail](https://user-images.githubusercontent.com/1186631/37345241-c05bba92-26cc-11e8-8de0-cc515991a836.png)

Clicking on compute resource name takes me to compute attrs edit page. I'd expect the breadcrumbs to have 3 levels there:
`Compute profiles >> 1 - Small >> Edit compute profile on oVIrt (oVirt)`
![compute_profile_edit_attributes](https://user-images.githubusercontent.com/1186631/37345242-c1ddd8f0-26cc-11e8-91c6-b93e69c71116.png)

Also the first level link takes you to the same url on this page.

tstrachota commented:
Similar issues are on the Images page.
Could it be nested in `Compute Resources >> [compute_resource.name] >> Create image` ?
![create_image](https://user-images.githubusercontent.com/1186631/37346261-27e2f66a-26cf-11e8-861f-2a1bc4a9337b.png)

The same issue with the first level link is here too.

amirfefer commented:
@ohadlevy , @tstrachota Thanks for your great feedbak!
I fixed title styles, and pages that required a specific breadcrumbs.

Could you have another look ?

tstrachota commented on app/views/images/new.html.erb:
```diff
@@ -1,3 +1,8 @@
+      [
+        {caption: @compute_resource.to_s, url: (url_for(compute_resource_path(@compute_resource)) if authorized_for(hash_for_compute_resources_path))},
+        {caption: _("Images"), url: (url_for(compute_resource_images_path(@compute_resource)) )},
+        {caption: _("Create image")}
```
Link to compute resource index is missing here.

Compare to images edit page:
`oVIrt (oVirt) >> Images >> Create image`
vs.
`Compute resources >> oVIrt (oVirt) >> Images >> Edit ...`

tstrachota commented on app/views/filters/edit.html.erb:
```diff
@@ -1,3 +1,3 @@
-<% title _("Edit Filter") %>
+<% title _("Edit %s") % @filter.to_s %>
```
Could you add breadcrumbs for `Roles >> %{role_name} filters >> Edit filter` here?
Users come there from the roles page and switching the root breadcrumb to "Filters" feels unnatural. The same for the create page, please.

tstrachota commented on app/views/compute_attributes/new.html.erb:
```diff
@@ -1,3 +1,9 @@
+        {caption: _("Compute Resources"), url: (url_for(compute_resources_path) if authorized_for(hash_for_compute_resources_path))},
+        {caption: _("Compute Profiles"), url: (url_for(compute_profiles_path) if authorized_for(hash_for_compute_profiles_path))},
+        {caption: @set.compute_resource.to_label, url: (compute_resource_path(@set.compute_resource) if authorized_for(hash_for_compute_resource_path(@set.compute_resource)))},
+        {caption: _('New compute profile')}
```
This one is tricky. Users can get there from both compute resource detail and compute profile detail. I've been thinking about it and most natural combination I could find is:
`Compute Resources >> %{compute_resource_label} >> Compute Profiles >> New %{profile_name}`
eg.:
`Compute Resources >> tstracho-laptop (Libvirt) >> Compute Profiles >> New 2-Medium`
Where "Compute Profiles" points to `/compute_profiles/`.

Similarly for the edit page:
`Compute Resources >> %{compute_resource_label} >> Compute Profiles >> Edit %{profile_name}`

This way it feels quite natural when coming from the compute resource detail. When coming from compute profiles page the order of breadcrumbs is swapped compared to what I'd expect but this is something I can live with. It will also correspond with breadcrumbs for images. What do you think?

tstrachota reviewed (changes requested):
Great update @amirfefer, thanks. Broken links are fixed now.

**Things that still need to be fixed:**
1. There are 3 pages where the breadcrumbs still need to be tuned (see my inline comments).

2. I'm getting warning in the console:
```
Warning: Invalid DOM property `class`. Did you mean `className`?
    in h1 (created by ForemanBreadcrumb)
    in div (created by ForemanBreadcrumb)
    in ForemanBreadcrumb
    in Provider
```

amirfefer commented:
@tstrachota thanks again.
 I’ll appreciate it if you can have another look.
with these changes there are some broken integration tests though.

ohadlevy commented on webpack/assets/javascripts/react_app/components/Breadcrumb/index.js:
```diff
@@ -0,0 +1,27 @@
+const ForemanBreadcrumb = ({ data }) => {
+  if (data.menu.length === 1) {
+    return (
+      <div class="row form-group">
```
please use className

ohadlevy commented:
@Rohoover i don't like the space between the buttons and the breadcrumbs above it, it seems like a lot of wasted space, does it make sense to have them in the same height? 
![image](https://user-images.githubusercontent.com/20167/37570019-0e72fd1a-2af3-11e8-81c8-c30131483976.png)


nitpick: maybe the image should align with the text size?

ohadlevy commented:
when clicking on a host audits, I was expecting to see breadcrumbs as well? e.g.
![image](https://user-images.githubusercontent.com/20167/37570045-491a98b0-2af3-11e8-8fde-032d649429e6.png)

ohadlevy commented:
similar issue with reports (when clicking on a host configuration report) 
![image](https://user-images.githubusercontent.com/20167/37570066-9a206442-2af3-11e8-8f89-2d42fbc3374c.png)

serenamarie125 commented:
>  i don't like the space between the buttons and the breadcrumbs above it, it seems like a lot of wasted space, does it make sense to have them in the same height?

In this case, according to PF, the buttons should share the same "bar" as the breadcrumb

Rohoover reviewed (commented):
@serenamarie125 @ohadlevy 

Based on dev feedback, further up in the thread we discussed using the Horizontal line to create a consistent space for the breadcrumbs. This feature is  outlined in the PF documentation:
"Divider (optional): A horizontal divider may be present below the breadcrumbs depending on the page layout used inside an application. If this format is used, it should be used consistently throughout the application.
"
As per the documentation, if we're going to use this visual element, it should be used across pages. 

Regardless of line or not, bringing the breadcrumbs in line with those buttons would be very inconsistent with how we've been handling it on every other page, where it is not in line with any page actions. 

Also, I've seen a screen of the shared breadcrumb and page row and there is risk that long named hosts would collide with the buttons.

amirfefer commented:
@ohadlevy  something like this?
![breadcrumbs-importent](https://user-images.githubusercontent.com/11807069/37597810-1bb495ec-2b89-11e8-8c61-71efde3f8979.png)

It doesn't feel like the pf design, plus the breadcrumbs may break like this:
![fireshot capture 77 - - http___localhost_3000_compute_resources_1-libvirt_images_1_edit](https://user-images.githubusercontent.com/11807069/37598039-c2158874-2b89-11e8-9ad8-523aaf74674d.png)

 inconsistent  with other pages:
![fireshot capture 78 - hosts - http___localhost_3000_hosts](https://user-images.githubusercontent.com/11807069/37598548-2cedf644-2b8b-11e8-9fde-4c02cd48e40d.png)

waldenraines commented:
> In this case, according to PF, the buttons should share the same "bar" as the breadcrumb

I don't see that in [the documentation](http://www.patternfly.org/pattern-library/navigation/breadcrumbs/) or in the [navigation guidelines](http://www.patternfly.org/pattern-library/navigation/guidelines/).  Where is this documented on the patternfly site?  

I think we should have the buttons in a separate row for consistency and we should revisit this in the future once the breadcrumb and button placement best practices are better documented.

Rohoover commented:
@waldenraines 
I checked with the PF UX Strategist and was not able to locate any documentation on this issue. 

I believe combing breadcrumbs with action button will lead us down a path fraught with exceptions:
1) What will we do when we have the search bar?
2) What will we do when we have too many action buttons?
3) What will we do when the breadcrumb title becomes too long?

Since this is an item of navigational context as well as an acting page title, it's best positioned at the top of the page to "set the stage" for all things below it. By putting it in line with actions, we are conflating it a bit with action buttons.

Below is a dedicated breadcrumb title area with horizontal line (to separate the space and as a visual element):

![screen shot 2018-03-19 at 10 41 48 am](https://user-images.githubusercontent.com/18266453/37602342-85e52638-2b62-11e8-971e-e990105da5d3.png)

Using this layout, I believe, exceptions would be far more rare.

ohadlevy commented:
I tend to agree this might be a less than ideal path forward, but I can't
say I like the current (see my earlier attachment) solution either, would
it be better to add the horizontal line or some other way to make the
spacing a bit less distracting? (or maybe its just me..)

On Mon, Mar 19, 2018 at 4:45 PM, Roxanne Hoover <notifications@github.com>
wrote:

> @waldenraines <https://github.com/waldenraines>
> I checked with the PF UX Strategist and was not able to locate any
> documentation on this issue.
>
> I believe combing breadcrumbs with action button will lead us down a path
> fraught with exceptions:
>
>    1. What will we do when we have the search bar?
>    2. What will we do when we have too many action buttons?
>    3. What will we do when the breadcrumb title becomes too long?
>
> Since this is an item of navigational context as well as an acting page
> title, it's best positioned at the top of the page to "set the stage" for
> all things below it. By putting it in line with actions, we are conflating
> it a bit with action buttons.
>
> Below is a dedicated breadcrumb title area with horizontal line (to
> separate the space and as a visual element):
>
> [image: screen shot 2018-03-19 at 10 41 48 am]
> <https://user-images.githubusercontent.com/18266453/37602342-85e52638-2b62-11e8-971e-e990105da5d3.png>
>
> Using this layout, I believe, exceptions would be far more rare.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/theforeman/foreman/pull/5167#issuecomment-374237367>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABOxz3YN-LkcnOEZFkyXXyM0tcuXbwuks5tf8SOgaJpZM4RZvfs>
> .
>

Rohoover commented:
@ohadlevy 
I share your concern about the space but see it more as small visual issue and not necessitating breaking of consistency (and pf). If anything I think there is a bigger issue with the host detail page. 

I think designing for the majority of pages vs. the one page (that probably itself needs to be redesigned - host merge addresses this) would be a better approach.

amirfefer commented:
@Rohoover Thanks for your feedback
I added a horizontal line beneath the breadcrumbs:

![compute_with_line](https://user-images.githubusercontent.com/11807069/37606695-c2f5111e-2b9e-11e8-8f96-4daaa16cdae7.png)
@ohadlevy I inserted specific breadcrumbs in reports and audits pages from a host.

serenamarie125 commented:
> better to add the horizontal line

@Rohoover based on the [design doc for breadcrumbs](http://www.patternfly.org/pattern-library/navigation/breadcrumbs/#design) it does show a horizontal line under, what do you think about that for a solution?  We are doing this in some other products already.

serenamarie125 commented:
> I checked with the PF UX Strategist and was not able to locate any documentation on this issue.

@Rohoover let's chat with Matt C, as I think there's some work in the backlog around the topic

amirfefer commented:
@Rohoover, which horizontal line looks better?
![horizontal-1](https://user-images.githubusercontent.com/11807069/37646563-aef3d9cc-2c32-11e8-95f3-b077534b15d0.png)

![horizontal-1](https://user-images.githubusercontent.com/11807069/37646664-eaf239a0-2c32-11e8-8771-bfddcdd17989.png)

tstrachota commented:
+1 for the full-width, plus the breadcrumbs should imho be vertically centered.

tstrachota commented on app/views/audits/index.html.erb:
```diff
@@ -1,3 +1,13 @@
+     [
+       {caption: _("Hosts"), url: (url_for(hosts_path) if authorized_for(hash_for_hosts_path))},
+       {caption: @host.name, url: (host_path(@host) if authorized_for(hash_for_host_path(@host)))},
+       {caption: _('Audits'), url: url_for(audits_path)},
```
It should imho be just `Hosts > %{host_name} > Audits`:
```ruby
[
  {caption: _("Hosts"), url: (url_for(hosts_path) if authorized_for(hash_for_hosts_path))},
  {caption: @host.name, url: (host_path(@host) if authorized_for(hash_for_host_path(@host)))},
  {caption: _('Audits')},
]
```

Having a link to listing all audits nested under hosts doesn't make much sense to me.

tstrachota commented on app/views/config_reports/index.html.erb:
```diff
@@ -1,3 +1,13 @@
+      {caption: _("Hosts"), url: (url_for(hosts_path) if authorized_for(hash_for_hosts_path))},
+      {caption: @host.name, url: (host_path(@host) if authorized_for(hash_for_host_path(@host)))},
+      {caption: _('Reports'), url: url_for(config_reports_path)},
+      {caption: _('Reposrt for %s ') % host.name}
```
Same here, I find `Hosts > %{host_name} > Reports` more intuitive.

tstrachota commented:
Compute profile details and host detail pages throw a warning:
```
Warning: Received `true` for a non-boolean attribute `title`.
If you want to write it to the DOM, pass a string instead: title="true" or title={value.toString()}.
    in ol (created by Breadcrumb)
    in Breadcrumb (created by ForemanBreadcrumb)
    in div (created by ForemanBreadcrumb)
    in ForemanBreadcrumb
    in Provider
```

tstrachota commented:
Could we re-visit the breadcrumbs structure for hosts and proxies? I really find it quite annoying that I can't use breadcrumbs to get from the edit page back to the detail. Is it just me? :)

Rohoover reviewed (commented):
@amirfefer +1 to full width

I think the summary of this conversation is that the best approach currently supported by PF that would provide the most consistent experience, is to include the horizontal line (see my above comments and mockup). 

This would reduce instances of exceptions because it fits almost all use cases (I actually can't think of one that it doesn't), give the breadcrumbs their own dedicated space (especially if they're quite long), and offers them proper placement at the top of a page which serves their purpose as both a navigational and contextual feature.

This also helps visually differentiate the space as this was a previous concern which is what sparked the grey background discussion earlier.

Things to consider for the future are instances in which unbalanced white space is present (host details). This will hopefully be addressed in the future with a redesign.

amirfefer commented:
@tstrachota keep in mind that we already have `edit` button in the `title_actions` (host details page)
![fireshot capture 88 - damon-rouselle test - http___localhost_3000_hosts_damon-rouselle test](https://user-images.githubusercontent.com/11807069/37669096-abb2884e-2c6e-11e8-90cd-9b7e4c283592.png)
is this the right location ? should we move it to the breadcrumbs navigation? I'm not sure.

tstrachota commented:
@amirfefer the buttons on the detail page are totally fine. I'm just suggesting to change breadcrumbs on the host (and proxy) edit page to `Hosts > %{host_name} > Edit  %{host_name}` so that breadcrumbs provide link to do the step back.
Otherwise it's "click on all hosts" and "click on the host again"

ohadlevy commented:
Would adding the links in the action area will be better for consistency
usage cases?

On Tue, Mar 20, 2018, 7:05 PM Tomas Strachota <notifications@github.com>
wrote:

> @amirfefer <https://github.com/amirfefer> the buttons on the detail page
> are totally fine. I'm just suggesting to change breadcrumbs on the host
> (and proxy) edit page to Hosts > %{host_name} > Edit %{host_name} so that
> breadcrumbs provide link to do the step back.
> Otherwise it's "click on all hosts" and "click on the host again"
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/theforeman/foreman/pull/5167#issuecomment-374677839>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABOx8pF5ZRVFjRyv18M46LhD1eubDwEks5tgTbUgaJpZM4RZvfs>
> .
>

tstrachota commented:
I'd expect it to be in the breadcrumbs. Action buttons is different location. But if it's just me and everyone else feels it's ok as it is now, just ignore my comment:)
We can always tune it later if it turns out people have different preferences.

tstrachota commented:
I'm still seeing the warning I mentioned here: https://github.com/theforeman/foreman/pull/5167#issuecomment-374557787

When this is fixed, I think we're good to merge.

amirfefer commented:
@tstrachota  could you be more specific? when this warning appears exactly?

Rohoover reviewed (commented):
@tstrachota 
I see what you are saying and I don't disagree. 

The edit button essentially takes you to a new page, so from a "rules" perspective, it would make sense to indicate that in the breadcrumb. I would just ask that we ensure that we continue with this principle on other pages. 

Also agreed that if we find this to be ineffective, we can opt to change later.

tstrachota commented:
> when this warning appears exactly?

Please disregard, it was issue on my side. Everything is ok after I updated npm packages. 
:+1:

ohadlevy commented on app/controllers/audits_controller.rb:
```diff
@@ -5,6 +5,7 @@ class AuditsController < ApplicationController
 
   def index
     @audits = resource_base_search_and_page.with_taxonomy_scope.preload(:user)
+    @host = resource_finder(Host.authorized(:view_hosts), params[:host_id]) if params[:host_id]
```
please remove the back btn from the audits page 
![image](https://user-images.githubusercontent.com/20167/37873609-ca076f90-3028-11e8-8c02-4ca5f2d03474.png)

ohadlevy commented on app/controllers/config_reports_controller.rb:
```diff
@@ -7,6 +7,7 @@ class ConfigReportsController < ApplicationController
   def index
     respond_to do |format|
       format.html do
+        @host = resource_finder(Host.authorized(:view_hosts), params[:host_id]) if params[:host_id]
```
similar issue here (with back btn) 
![image](https://user-images.githubusercontent.com/20167/37873616-ebf64130-3028-11e8-8909-0b3674f5028a.png)

but please note the difference, the breadcrumbs links points to: /config_reports while the back btn points to /config_reports?search=eventful+%3D+true

please remove the back btn and change the url for the index.

Also - does it make sense to make the breadcrumbs link to the host details page? (that could eliminate the host details btn as well).

ohadlevy commented:
@amirfefer this looks pretty ready for me, with the minor comment about back buttons, please have another look and hopefully we can merge this soon.

ohadlevy commented:
@amirfefer also, not sure if test failures are related or not...

tstrachota commented:
:+1: to merge from my side once the tests are green.

I think that to be 100% visually correct we'd need to add some more margin from top to center the breadcrumbs in their row. The patternfly design has them also vertically centered.

current state:
![screenshot from 2018-03-27 15-13-25](https://user-images.githubusercontent.com/1186631/37970174-04edba00-31d3-11e8-939e-d732a173b303.png)
vs. centered:
![screenshot from 2018-03-27 15-14-10](https://user-images.githubusercontent.com/1186631/37970178-066a2b20-31d3-11e8-9b8b-10283e7aae2e.png)

But I'm good to merge it without that. It should be easy to tune later.

amirfefer commented:
[test hound]

amirfefer commented:
@ohadlevy  - I fixed tests failures

ares commented:
:clap:

waldenraines commented:
👍 🎉

tstrachota commented:
Great news! :clap:
